actioncable/lib/action_cable/connection/base.rb

```diff
@@ -120,19 +128,32 @@ def send_async(method, *arguments)
         worker_pool.async_invoke(self, method, *arguments)
       end
 
-      # Return a basic hash of statistics for the connection keyed with <tt>identifier</tt>, <tt>started_at</tt>, <tt>subscriptions</tt>, and <tt>request_id</tt>.
+      # Return a basic hash of statistics for the connection keyed with <tt>identifier</tt>, <tt>started_at</tt>,
+      # <tt>subscriptions</tt>, <tt>request_id</tt>, and <tt>last_message_received_at</tt>.
       # This can be returned by a health check against the connection.
       def statistics
         {
           identifier: connection_identifier,
           started_at: @started_at,
           subscriptions: subscriptions.identifiers,
-          request_id: @env["action_dispatch.request_id"]
+          request_id: @env["action_dispatch.request_id"],
+          last_message_received_at: @last_message_received_at
         }
       end
 
       def beat
-        transmit type: ActionCable::INTERNAL[:message_types][:ping], message: Time.now.to_i
+        unless expects_pong_response_to_heartbeat?
+          transmit type: ActionCable::INTERNAL[:message_types][:ping], message: Time.now.to_i
+          return
+        end
+
+        if half_open?
+          subscriptions.unsubscribe_from_all
+          close(reason: ::ActionCable::INTERNAL[:disconnect_reasons][:heartbeat_timeout])
+          return server.remove_connection(self)
```

Comment:
I'm not sure if this is the proper way to fully close a connection. Some input would be much appreciated.

Reply:
You are right, calling `#close` is enough.
Client socket defines `#begin_close` as it's close callback, which calls `#finalize_close`, which then calls `#on_close`.

When I first wrote this PR I was playing around with this and it seems like the subscriptions and connection were left in tact. I'll double check this to make sure, but for now I changed the code to just call `#close`.